File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        if(s.empty()) return "";
        int start = 0 , len = 0;
        for(int i=0; i<n; i++){
            int l  = i , r = i;
            while(r <n-1 && s[r] == s[r+1]) r++;
            i = r;
            while(l>0 && r<n-1 && s[l-1] == s[r+1]){
                l--;
                r++;
            }
            if(r-l+1 > len){
                len = r-l+1;
                start = l;
            }
        }
        return s.substr(start , len);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        String t = "#"+ String.join("#" , s.split("")) + "#";
        int n = t.length();
        int C = 0, R =0;
        int[] p = new int[n];
        
        for(int i=1; i<n-1; i++){
            int mirror = 2*C - i;
            if(i<R){
                p[i] = Math.min(R-i , p[mirror]);
            }
            while(i+p[i]+1 < n && i-p[i]-1 >=0 && t.charAt(i-p[i]-1) == t.charAt(i+p[i] + 1) ){
                    p[i]++;
            }
            if(i+p[i] > R){
                C = i;
                R = i+p[i];
                
            }
        }
        
        int maxLen = 0 , centerIdx = 0;
        for(int i=1; i<n-1; i++){
            if(p[i] > maxLen){
                centerIdx = i;
                maxLen = p[i];
            }
        }
        int start = (centerIdx - maxLen)/2;
        return s.substring(start , start + maxLen);
    }
}